src/main/java/org/spongepowered/common/mixin/core/inventory/ContainerMixin.java

```diff
@@ -431,26 +425,43 @@ private void beforeSlotChangedCraftingGrid(final InventoryCraftResult output, fi
         final ItemStackSnapshot repl = ItemStackUtil.snapshotOf(output.getStackInSlot(index));
 
         final SlotAdapter slot = this.impl$getAdapters().get(index);
-        this.impl$capturedCraftPreviewTransactions.add(new SlotTransaction(slot, orig, repl));
+        if (this.impl$capturedCraftPreviewTransaction == null) {
+            this.impl$capturedCraftPreviewTransaction = new SlotTransaction(slot, orig, repl);
+        } else if (this.impl$capturedCraftPreviewTransaction.getSlot().equals(slot)
```

Comment:
Off the top of my head, do `SlotAdapter`s implement `equals`?

Reply:
Not currently. It seems to be an easy implementation - it just requires checking the underlying inventory and the index - so I'll do that.